### ISSUE TITLE
Jinn "crashes/dies" during multiple Pay Request run

### ISSUE THREAD

**Issue by jnbarchan** [hezmondo/Jinn#266]
When doing an **MPR Selected Run**, outputting to PDF, of a "large" number of rents, Jinn slows down and eventually completely "freezes" (may be crash, may be die, may be freeze).  You usually have to completely reboot machine to unfreeze.  This behaviour has been observed under both Linux & Windows.

It is enough to invoke this via **Preview >>**, you do not have to **Run >>**.  The precise number of rents to run to exhibit behaviour depends on free memory of machine --- several hundred to several thousand.

As code stands at the moment, even doing multiple small runs will not help --- as the total number of runs increases, the "freeze" will occur later on.  Only exiting Jinn and restarting it will resolve the issue.

**Comment by jnbarchan** [comment]
Inspecting via Linux 
```
watch -n 1 free mem
watch -n 1 ps -C QtWebEngineProc
top -o %MEM
```
(presumably something similar from **Task manager** under Windows) reveals ever increasing memory is allocated to Jinn/QtWebEngineProc processes.  It never gets freed.

**Comment by jnbarchan** [comment]
In `writeMail()` code loops through each rent calling
```
preview = HtmlPreviewDialog(...)
```
This is used to create an invisible, non-interactive dialog holding a `QWebEngineView` to do the print to PDF from.  These dialogs/webengine views remain in memory, they do not get disposed.  As each takes up several MB, eventually the machine runs out of memory, causing the "freeze".

At least in the case where the `preview` is used to repeatedly create non-interactive dialogs, it needs deleting.  `preview.deleteLater()` improves the situation, in that the memory used gets freed *but only when return is made to the top-level Qt event loop*, i.e. at the *end* of the run.  (This is how Qt works --- for unknown code reasons, only the top-level Qt event loop deals with `deleteLater()`s, putting in, say, an explicit `processEvents()` will not cause memory release.)  But *during* the run the memory still does not get released, so user would be limited to having to do multiple, small runs.

I asked about this behaviour in **pyqt mailing list**.  One suggestion was reparenting the dialog to have no parent, but I found this did not help.  The other suggestion was

> If you need to parent the dialogs, another option is to use "sip.delete(dlg)" instead of "dlg.deleteLater()", which will trigger immediate deletion of the dialog instead of scheduling it for late
> 
> If that doesn't work [dialog re-parenting], then sip.delete should be safe so long as there are no pending events for the dialog being deleted.  The Qt documentation for the QObject destructor has a little info about this: https://doc.qt.io/qt-5/qobject.html#dtor.QObject.  (sip.delete corresponds to a C++ delete, which would immediately trigger the destructor.)  You could call processEvents before calling sip.delete, since that should take care of any pending events.

I am going to (have to) go for `sip.delete()`, as that does release the memory immediately and is the only way forward given our code....

**Comment by jnbarchan** [comment]
To test: suggest before downloading this fix, do a run/preview with a certain number of rents (can be fewer than required to "freeze") and watch what (Windows) **Task Manager** shows about used/available/free memory.  You should see memory getting chewed up, until run completes.  Then try new Jinn version with this fix, and verify memory released as run progresses instead of at end.

**Comment by LaurenceLeo** [comment]
Hey J,
Not sure if I am providing you with anything helpful here. 
I realised earlier that in WriteMail  I neglected to put a done/accept/reject for the situation where multiple rents are used. This meant that the JDialogs were never being destroyed (my understanding from testing). After adding a "preview.done()" at the end of the multiple rent run. I found that all of the QT QtWebEngineProcesses closed after WriteMail was complete. 

They still ballooned during a rent run though, I believed this was due to what you mentioned above. I was planning on taking the QtWebEngine parts out of the JDialog and working them into HtmlUtils. This way the QtWebEngine would not be called from within HtmlPreviewDialog and we should be able to avoid having to wait for them to be garbage collected along with the JDialog. I was not entirely sure this would work. 

Not sure if any of this information is helpful but if there is a chance it helps
